src/lib.rs

```diff
@@ -0,0 +1,217 @@
+    let args = match cmd_args {
+        None => CmdArgs::from_args(),
+        Some(cmd_args) => CmdArgs::from_iter_safe(cmd_args).map_err(|err| err.to_string())?,
+    };
+
+    let vault_bin_path = get_vault_bin_path(args.vault_path)?;
+    let msg = format!(
+        "Launching with vault executable from: {}",
+        vault_bin_path.display()
+    );
+    if args.verbosity > 0 {
+        println!("{}", msg);
+    }
+    debug!("{}", msg);
+
+    let msg = format!("Network size: {} vaults", args.num_vaults);
+    if args.verbosity > 0 {
+        println!("{}", msg);
+    }
+    debug!("{}", msg);
+
+    let mut common_args: Vec<&str> = vec![];
+
+    let mut verbosity = String::from("-");
+    if args.vaults_verbosity > 0 {
+        for _ in 0..args.vaults_verbosity {
+            verbosity.push('v');
+        }
+        common_args.push(&verbosity);
+    }
+
+    // Construct genesis vault's command arguments
+    let mut genesis_vault_args = common_args.clone();
+    genesis_vault_args.push("--first");
+    let genesis_vault_dir = &args.vaults_dir.join("safe-vault-genesis");
+    let genesis_vault_dir_str = genesis_vault_dir.display().to_string();
+    genesis_vault_args.push("--root-dir");
+    genesis_vault_args.push(&genesis_vault_dir_str);
+    genesis_vault_args.push("--log-dir");
+    genesis_vault_args.push(&genesis_vault_dir_str);
+
+    // Let's launch genesis vault now
+    let msg = "Launching genesis vault (#1)...";
+    if args.verbosity > 0 {
+        println!("{}", msg);
+    }
+    debug!("{}", msg);
+    run_vault_cmd(&vault_bin_path, &genesis_vault_args, args.verbosity)?;
+
+    // Get port number of genesis vault to pass it as hard-coded contact to the other vaults
+    let interval_duration = Duration::from_secs(args.interval);
+    thread::sleep(interval_duration);
+    let genesis_contant_info = grep_connection_info(&genesis_vault_dir.join("safe_vault.log"))?;
+    let msg = format!("Genesis vault contact info: {}", genesis_contant_info);
+    if args.verbosity > 0 {
+        println!("{}", msg);
+    }
+    debug!("{}", msg);
+
+    // We can now run the rest of the vaults
+    for i in 2..args.num_vaults + 1 {
+        // Construct current vault's command arguments
+        let mut current_vault_args = common_args.clone();
+        let vault_dir = &args
+            .vaults_dir
+            .join(&format!("safe-vault-{}", i))
+            .display()
+            .to_string();
+
+        current_vault_args.push("--root-dir");
+        current_vault_args.push(vault_dir);
+        current_vault_args.push("--log-dir");
+        current_vault_args.push(vault_dir);
+        current_vault_args.push("--hard-coded-contacts");
+        current_vault_args.push(&genesis_contant_info);
+
+        let msg = format!("Launching vault #{}...", i);
+        if args.verbosity > 0 {
+            println!("{}", msg);
+        }
+        debug!("{}", msg);
+        run_vault_cmd(&vault_bin_path, &current_vault_args, args.verbosity)?;
```

Comment:
This code seems to be repeated between the genesis and other vaults. Maybe we can refactor this and have a single function to do this, something like `launch_vault(is_genesis: bool)`.

Reply:
I see we have a `run_vault_cmd` function but that doesn't involve parsing the args and generating a final command.

Reply:
`run_vault_cmd` expects the args to pass to vault. I'll see if I can put the args builder in a helper function as you say